### ISSUE TITLE
Failure to Sparsify Forced Linear System with Large Parameter Differences

### ISSUE THREAD

**Issue by JGPrior** [dynamicslab/pysindy#99]
Hello to whom it may concern,

Whilst I don't have a specific and identifiable issue with PySINDy functionality itself, I am unfortunately experiencing some issues with creating a sparse and representative model of a system that I perceive to be  well within the bounds of SINDy's capability. The system at question is a simple spring-damper quarter car suspension model that is linear in four states. The state space representation of the system is:

![Screenshot 2020-12-03 112407](https://user-images.githubusercontent.com/62751537/101010499-3a842680-355a-11eb-9489-24bc99d52be6.jpg)

I have simulated this system in MATLAB/SIMULINK (tried to solve this problem with the original MATLAB implementation before moving to Python) by providing a step input in _u_ and using a fixed step (t = 0.001) ODE3 solver. My understanding of linear systems is that such a step input should be sufficient to characterise them.

I have subsequently moved to Python and imported the MATLAB array data using:

`from scipy.io import loadmat`
`matcontents = loadmat('sim.mat')`
`simdata = matcontents['simdata']`
`u_train = simdata[:,4]`
`x_train = simdata[:,:4]`

where the first four columns of `simdata` are the state variables and fifth is the input. I can also provide the original sim.mat if required. I have created a poly library up to 2nd degree + bias, and my default optimiser choice has been STLSQ, threshold = 0.5 and alpha = 0.

The resulting model is:
![Screenshot 2020-12-03 114926](https://user-images.githubusercontent.com/62751537/101014399-bd5ab080-355d-11eb-9419-f87057acb381.jpg)

It should be clear that both `x1'` and `x3'` are correctly identified. However, `x4'`, has failed to sparsify to the governing equation, despite having near perfectly captured the correct dynamics in the linear measurements. Addtionally, the solution for `x2'` is both completely incorrectly identified and not adequately sparsified. This result implies that the data characterising my system is correct (i.e. no errors in the data-generating model), but that there is something wrong with my attempt to create a data-driven model of it.

I have attempted myriad potential fixes, including:

- Introducing a non-zero alpha value to the optimiser. This has produced an even less sparse solution in the problem states.
- Normalising the measurement library. I thought this would instantly fix the problem but seems to have no effect on the output.
- Thresholding incredibly small values (<10-5) of my states in the training data to 0 to reduce the range of values in the feature space. Even when I perform this and normalise there seems to be no effect.
- Altering my excitation signal design from a step input to white noise. This also seems to have little effect on the result.
- I have even attempted to increase the threshold of STLSQ to >1 (a value which regularises out all parameters in 2 out of 4 states) to attempt to sparsify `x2'`, which has large parameter values. However, even this fails to converge to a true solution for any threshold values, regardless of how high I set it.
- I have attempted to use SR3 with default values and some minor tuning, but the results appears to suffer from similar lack of sparsity. I will admit, however, to not having performed a full cross-validation to ascertain hyper-parameters for this optimiser, due to my lack of familiarity.
- I've even thrown an Elastic Net at the problem, which has promoted better sparsity but produced some laughably inaccurate parameter values.

Beyond performing full hyper-parameter CV over multiple optimisers, I am at somewhat of a loss with how to move forward. This system was meant to be a demo case before moving onto a non-linear spring-damper version and subsequently a Discrepancy Model application inspired by the work of the Washington research team, but if I cannot get SINDy to work in this simple dynamic system with basic attributes I may have to abandon my work with it. Is the optimiser struggling because of the 10^3 order of magnitude difference between the largest and smallest parameter of the model, or are there more obvious faults in my approach to using SINDy that I havent yet realised?

Many thanks in advance for your assistance and kind regards,
Jack

**Comment by akaptano** [comment]
I tend to think you are right that the 10^3 order of magnitude difference in the parameters is the issue here. Certainly it is preventing you from systematically turning the threshold up and down, since the identified coefficients are very different for each ODE. I am not sure if this will fix the issue, or you have tried it already, but you could try normalizing each of the x_1,...,x_4 to the unit ball. So x_1 -> x_1/max(|x_1|) and so on. If the difference in parameter sizes in representative of a mix of "slow" and "fast" dynamics, take a look at https://arxiv.org/pdf/2006.00940.pdf. 

If you would like an easier but perhaps more ad-hoc approach, look at the SR3 branch. There we have SR3 implemented with initial guesses and variable thresholding. That means for each x_i you can set a different threshold lambda_i. You can even set higher thresholds only on quadratic terms. 

Also, as a first check to your identified model that you posted above, could you attach the plots of "true" and "identified" Xdot? It's possible that even though it does not capture the true ODE system well, it is still approximating Xdot pretty well. A picture like this could help inform what is going wrong here.

**Comment by JGPrior** [comment]
Hi Alan, thanks for the rapid response.

Firstly, for the plots of predicted vs computed:

![image](https://user-images.githubusercontent.com/62751537/101066733-8a7ce080-358e-11eb-9661-bfb4c22b36cd.png)

As you said, the model seems well fit to the data, it just doesn't capture the original ODE. Just as a side note: the above plots are for the training data. I haven't generated a validation set yet, but I'm sure this would demonstrate some degree of overfitting, particularly in x2_dot. 

I've also attempted to fit SINDy to data normalised in a unit hypersphere, as you suggested:

![image](https://user-images.githubusercontent.com/62751537/101068858-27d91400-3591-11eb-86c9-844d92ff62b0.png)

Again, when I plot the derivatives for a model based on this data the fit is good, arguably better than originally in x2_dot:

![image](https://user-images.githubusercontent.com/62751537/101068991-55be5880-3591-11eb-8753-b37672610b7f.png)

But the ODEs are somewhat baffling...

![image](https://user-images.githubusercontent.com/62751537/101069062-6c64af80-3591-11eb-973e-4b90b787587d.png)

This result comes using a threshold of 10, but there are a lot of remaining parameters that exceed that threshold somehow, even with alpha set to 0... I Have a feeling that I've messed something up here but it's hard to tell. I guess this is also in part because even a well-sparsified result is going to be fundamentally different to the original ODEs I used to generate the data, owing to the normalisation.

I'm definitely interested in the multi-scale DMD + SINDy solution you've proposed here. From the plots it's clear to see that my system has different speed dynamics, which would be entirely expected of a damped oscillating system such as a suspension model. However I may need to read over it a bit more before attempting, as I'm sure there are some differences in my application to some of the basic cases presented (forced, damped oscillations etc). Out of interest, is there a recommended repository for a DMD package?

Apologies for the long winded response, I'll be interested to know if this has provoked any further thoughts.

**Comment by akaptano** [comment]
Yeah the threshold not working could be a red flag that you're not using the module right. Could you post your code (at least the SINDy relevant part)? And get rid of the normalization for now so we can compare the coefficient values to the true model directly by eye.

**Comment by JGPrior** [comment]
Both the training data and the code:

[demo_Qcar.zip](https://github.com/dynamicslab/pysindy/files/5638675/demo_Qcar.zip)

**Comment by akaptano** [comment]
Okay, it looks like this works fine with poly_order = 1 (since this is a linear system). With poly_order = 2, it also fits Xdot very well, but the odeint(Xdot) is liable to blow up if you overfit with quadratic terms. Attached are my results for poly_order = 1, with nothing else changed (except I integrate the identified model to check the fit on x1,...,x4 and not just on x1dot,...,x4dot). Note that the model still does not agree with your ground truth, but notably it disagrees on precisely the locations where the coefficients are very large. For whatever reason, the fit of X is still very good -- the model cares little about the values where it is wrong. My guess is that you are not training on data for which the uncertainty in these coefficients can be reduced significantly (the "fast" dynamics are not resolved...?). I haven't looked at the physics so I'll leave that to you. Let me know if you have any more questions. 

[Qcar_x.pdf](https://github.com/dynamicslab/pysindy/files/5639079/Qcar_x.pdf)
[Qcar_xdot.pdf](https://github.com/dynamicslab/pysindy/files/5639080/Qcar_xdot.pdf)
<img width="628" alt="Screen Shot 2020-12-03 at 1 39 56 PM" src="https://user-images.githubusercontent.com/37751714/101091695-5e9d3300-356d-11eb-9a0a-ab6d5d69367b.png">

**Comment by briandesilva** [comment]
@akaptano, thanks for your work on this. I'll look into changing the default behavior of `SINDy.simulate`. Other people have reported similar problems in the past.

**Comment by briandesilva** [comment]
> Out of interest, is there a recommended repository for a DMD package?

[PyDMD](https://github.com/mathLab/PyDMD) is the best one that I'm aware of. It implements lots of different DMD variants and has good a good set of tutorials for getting started.

As for the issues with normalization, there's a bit of subtlety when it comes to normalizing the library terms. For example, if you use a `STLSQ` optimizer with a threshold of 0.5 and set `normalize` to true
```python
STLSQ(threshold=0.5, normalize=True)
```
you don't necessarily get an ODE with coefficients all larger than 0.5 back. The reason for this is because the normalization process boils down to
1. Compute the library features Theta(X) (each column corresponds to a different function);
2. Divide each column of Theta(X) by its norm, saving the norms for later. Let Theta*(X) be the resulting array;
3. Solve X' = Theta*(X) Xi using STLSQ;
4. Multiply the rows of Xi by the norms saved in step 2

All the nonzero values of Xi that are found in step 3 will be larger (in magnitude) than the threshold you specified, but after step 4, this property no longer holds.

**Comment by akaptano** [comment]
Thanks Brian, also @JGPrior, just so you are aware, the thresholding can be a bit subtle, especially with SR3, which does not hard-threshold. You can look in the published SR3 papers by Kathleen and see that sparsity is not enforced on Xi, but rather on a matrix W, and then there is an extra term in the optimization to approximately enforce W = Xi. The sparsity enforcement is "soft" or "relaxed" in this sense, and in fact the term promoting sparsity lambda*R(W), requires that lambda = threshold^2/nu, even though the L0 term (or L1 term) in R(W) is still chopping off terms in W > threshold. So changing threshold indeed changes where the L0 thresholding takes place (as desired), but it also scales the sparsity promotion term lambda*R(W) like O(lambda^2)
